pkg/resources/management.cattle.io/v3/globalrole/validator.go

```diff
@@ -61,26 +71,93 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
+	if err != nil {
+		return nil, fmt.Errorf("unable to resolve rules for new global role: %w", err)
+	}
+	err = auth.ConfirmNoEscalation(request, clusterRules, "", a.grbResolver)
+	if err != nil {
+		return admission.ResponseFailedEscalation(err.Error()), nil
+	}
+
 	// ensure all PolicyRules have at least one verb, otherwise RBAC controllers may encounter issues when creating Roles and ClusterRoles
 	for _, rule := range newGR.Rules {
 		if len(rule.Verbs) == 0 {
 			return admission.ResponseBadRequest("GlobalRole.Rules: PolicyRules must have at least one verb"), nil
 		}
 	}
 
-	response := &admissionv1.AdmissionResponse{}
-	auth.SetEscalationResponse(response, auth.ConfirmNoEscalation(request, newGR.Rules, "", a.resolver))
+	rules := a.grbResolver.GlobalRoleResolver.GlobalRulesFromRole(newGR)
 
-	return response, nil
+	err = auth.ConfirmNoEscalation(request, rules, "", a.resolver)
+	if err != nil {
+		return admission.ResponseFailedEscalation(err.Error()), nil
+	}
+
+	return admission.ResponseAllowed(), nil
+}
+
+// validateInheritedClusterRoles validates that new RoleTemplates specified by InheritedClusterRoles have a context of
+// cluster and are not locked. Does NOT check for user privilege escalation. May return a field.Error indicating the
+// source of the error.
+func (a *admitter) validateInheritedClusterRoles(oldGR *v3.GlobalRole, newGR *v3.GlobalRole, fieldPath *field.Path) error {
+	// fetch the old role templates as a map so that we can check which ones from newGR are new
+	oldRoleTemplates := map[string]struct{}{}
+	if oldGR != nil {
+		for _, oldRT := range oldGR.InheritedClusterRoles {
+			oldRoleTemplates[oldRT] = struct{}{}
+		}
+	}
+
+	var currentRoleTemplates []*v3.RoleTemplate
+	var err error
+	if newGR != nil {
+		currentRoleTemplates, err = a.grbResolver.GlobalRoleResolver.GetRoleTemplatesForGlobalRole(newGR)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				reason := fmt.Sprintf("unable to find all roleTemplates %s", err.Error())
+				return field.Invalid(fieldPath, "", reason)
+			}
+			return fmt.Errorf("unable to get roleTemplates for current version of GlobalRole %s: %w", oldGR.Name, err)
+		}
+	}
+
+	var newRoleTemplates []*v3.RoleTemplate
+	for _, currentRT := range currentRoleTemplates {
+		if _, ok := oldRoleTemplates[currentRT.Name]; !ok {
+			newRoleTemplates = append(newRoleTemplates, currentRT)
+		}
+	}
+
+	// if an RT is locked after the GR is created, we don't want to reject the request. But we also don't want
+	// users to add a locked RT as new permissions
+	for _, newRT := range newRoleTemplates {
+		if newRT.Context != roleTemplateClusterContext {
+			reason := fmt.Sprintf("unable to bind a roleTemplate with non-cluster context: %s", newRT.Context)
+			return field.Invalid(fieldPath, newRT.Name, reason)
+		} else if newRT.Locked {
+			reason := "unable to use locked roleTemplate"
+			return field.Invalid(fieldPath, newRT.Name, reason)
+		}
```

Comment:
I generally prefer the else if there since the two if statements are more connected in my mind than not connected (they both evaluate the validity of the target rt). But I can remove it - I would appreciate you pointing me to the style guide that indicates that this is preferred though.

Reply:
Done.